Create file to ingest power lines

JoelEldo opened the pull request:
Script that ingest power line data. I have also created a class called ingest_plines to do the same thing, but there are some kinks in that that need to be worked out before I can commit it here.

JoelEldo committed:
Create file to ingest power lines

Script that ingest power line data. I have also created a class called ingest_plines to do the same thing, but there are some kinks in that that need to be worked out before I can commit it here.

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+      geometry = feature['geometry']
+      info = feature['properties']
+      object = {}
+      object['shape'] = {'type':type, 'coordinates':geometry['coordinates']}                    # prepares the line into a multiline format acceptable by elasticsearch
+      object['type'] = type
```
Where does type get set?

InCadence replied:
Line 11:

type = 'multilinestring'

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+      'mappings': {
+         "properties": {
+
+            'location':{'type':'geo_shape'},
+            'OBJECTID':{'type':'integer'},
+            'ID':{'type':'integer'},
+            'TYPE':{'type':'text'},
+            'STATUS':{'type':'text'},
+            'NAICS_CODE':{'type':'integer'},
+            'NAICS_DESC':{'type':'text'},
+            'SOURCE':{'type':'text'},
+            'SOURCEDATE':{'type':'date',"format": "yyyy/MM/dd HH:mm:ss"},
+            'VAL_METHOD':{'type':'text'},
+            'VAL_DATE':{'type':'date',"format": "yyyy/MM/dd HH:mm:ss"},
+            'OWNER':{'type':'text'},
+            'VOLTAGE':{'type':'float'},
+            'VOLT_CLASS':{'type':'text'},
+            'INFERRED':{'type':'text'},
+            'SUB_1':{'type':'text'},
+            'SUB_2':{'type':'text'},
+            'SHAPE_Length':{'type':'float'}
```
All of these text fields should also have an associated keyword field.   The following is the mapping syntax:
"foo": {
    "type" "text",
    "fields": {
      "keyword": {
        "type": "keyword",
        "ignore_above": 256
      }
    }
  }

Also, this mapping file should be removed from being in-line in the code and read in from a configuration file.

JoelEldo replied:
I've committed a configuration file to the branch, and updated the code to read from said configuration file. I've also added the keyword field to all text fields in the configuration file.

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+            'SUB_1':{'type':'text'},
+            'SUB_2':{'type':'text'},
+            'SHAPE_Length':{'type':'float'}
+
+               }
+      }
+   }
+   es = Elasticsearch([{'host': 'localhost', 'port': 9200}],http_auth=('elastic','changeme'))   # initializes elasticsearch with authentication
+   if not es.indices.exists(index=index_name):                                                  # creates the index if it doesn't exist yet
+     es.indices.create(index=index_name,body=mapping)
+   return es
+
+
```
At some point, refactor this to be a generic create_connection routine that 
takes parameters from configuration files or arguements and can be used for any data and index.

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+
+
+def get_data():
+   return geojson.loads(requests.get(data_url, allow_redirects=True).content)["features"]       # gets data from url, then converts the geojson file into a dictionary 
```
Where is data_url set?
If this is specific to the electric power line data it should be something like get_electric_powerline_data

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+                                                                                                # of all the powerlines, then returns said dictionary
+
+def index_data(data_list, es):
+   for i in range(len(data_list)):                                                              #add data to index
+      es.index(index=index_name, id=data_list[i]['ID'], body=data_list[i])                      # id is the ID of the power line, unique to each power line
```
You should be able to do this as a bulk write.   You may have to manage chunk sizes.
I will send you some sample code.

bytemedwb reviewed (changes requested):
This is a good start.  Not the most efficient.
Please clearly set and show any constants (the type, the URL, etc.)

bytemedwb commented on ingest_power_lines.py:
```diff
@@ -0,0 +1,80 @@
+from elasticsearch import Elasticsearch                                                         # imports
+
+
+data_url = 'https://opendata.arcgis.com/datasets/70512b03fe994c6393107cc9946e5c22_0.geojson'    # sets up variables for future use
+if len(sys.argv) > 1:
+   data_url = sys.argv[1]                                                                       # option for user to choose their own data url
+index_name = 'powerlines'
+type = 'multilinestring'
+
```
Two things.
For constants I like the codeing convention that they are all UPPERCASE words.
Second don't intermix code and constants.   Move the argv check down into the main body.

JoelEldo replied:
I made a separate function that takes the user input. The only constant now is type, so that has been separated from the other variables.

JoelEldo committed:
Update ingest_power_lines.py

Updated it to a. bulk index, making it significantly faster, b. accept a mapping configuration file and c. accept user input for url and index name

JoelEldo committed:
Power Line Mapping Configuration File

JoelEldo committed:
Change some function names 

Change some function names to better represent what they do

JoelEldo committed:
Change an input prompt

I forgot to change "URL" to "file" when copy-pasting the input prompt.

JoelEldo committed:
Add create connection method, user input

Added create connection method and user input for elasticsearch connection.
